### PR TITLE
refactor compute_budget instruction processing as part of transaction sanitization, consolidate CU related limits into transaction meta data

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

`ComputeBudget::process_instructions()` is called multiple times at multiple locations during transaction's lifetime, causing: 
- redundant code. more places for bug
- higher code maintenance
- extra cost of calling `process_instructions()` are not accounted for

#### Aiming
- Main logic in `ComputeBudget::process_instructions()` doesn't really need `ComputeBudget` itself, it can be refactored out into its own module to "process" compute_budget instructions without `ComputeBudget`, so to live outside of `program-runtime`.
- making processing compute_budget ixs part of transaction sanitizing, since tx failed on these ixs will not be loaded nor executed;
- make limits set by these instructions available in sanitized transaction as meta data, helps leader scheduler, and fee calculator
- pave the way for TVL to replace compute budget instructions

#### Summary of Changes
- refactor compute_budget instructions process into its own module in `sdk`.
- remove cyclic dependency
- simplify `ComputeBudget` struct to its core functionality
- deprecate transaction_priority_details 
- refactored cost-model to be independent from compute_budget_intructions.
- remove fee_budget_limits from ComputeBudget module.
- expand transaction sanitization to include process compute_budget IXs to extract transaction_meta data

Fixes #

### CODE REVIEW

File: program-runtime/src/compute_budget.rs [hunk 1]
```diff
@@ -170,153 +143,24 @@ impl ComputeBudget {
...
-            loaded_accounts_data_size_limit: compute_budget.loaded_accounts_data_size_limit,
-            heap_cost: compute_budget.heap_cost,
-            compute_unit_limit: compute_budget.compute_unit_limit,
-            prioritization_fee: prioritization_fee_details.get_fee(),
-        }
+    pub fn new_from_transaction_meta(transaction_meta: &TransactionMeta) -> Self {
+        let mut compute_budget = ComputeBudget::new(u64::from(transaction_meta.compute_unit_limit));
+        compute_budget.heap_size = transaction_meta.updated_heap_bytes;
+        compute_budget
     }
```
Comment (taozhu-chicago) [inline program-runtime/src/compute_budget.rs:164]: improved `ComputeBudget` to be simply a struct has bunch of limits to be passed to runtime, nothing more.

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -4082,14 +4081,23 @@ impl Bank {
         message: &SanitizedMessage,
         lamports_per_signature: u64,
     ) -> u64 {
+        let transaction_meta = process_compute_budget_instruction(
+            message.program_instructions_iter(),
+            &self.feature_set,
+            Some(self.cluster_type()),
+        )
+        .unwrap_or_default();
```
Comment (taozhu-chicago) [inline runtime/src/bank.rs:4089]: transaction meta data are bound to `transaction`, not to `message`. Bank is given `SanitizedMessage` here to provide fee, forced one more ComputeBudget::process_instructinos() call here. It can be refactored away separately.

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -5123,37 +5140,22 @@ impl Bank {
...
+                                Measure::start("compute_budget_process_transaction_time");
+                            let transaction_meta = tx.get_transaction_meta();
+                            compute_budget_process_transaction_time.stop();
+                            saturating_add_assign!(
+                                timings
+                                    .execute_accessories
+                                    .compute_budget_process_transaction_us,
+                                compute_budget_process_transaction_time.as_us()
+                            );
+                            ComputeBudget::new_from_transaction_meta(&transaction_meta)
```
Comment (taozhu-chicago) [inline runtime/src/bank.rs:5157]: Saved one more `compute_budget.process_instructions(...)` call

File: runtime/src/prioritization_fee_cache.rs [hunk 4]
```diff
@@ -212,20 +209,16 @@ impl PrioritizationFeeCache {
                         continue;
                     }
 
-                    let round_compute_unit_price_enabled = false; // TODO: bank.feture_set.is_active(round_compute_unit_price)
-                    let priority_details = sanitized_transaction
-                        .get_transaction_priority_details(round_compute_unit_price_enabled);
+                    let transaction_meta = sanitized_transaction.get_transaction_meta();
```
Comment (taozhu-chicago) [inline runtime/src/prioritization_fee_cache.rs:212]: Another call of compute_budget.process_instructions(...) reduced

File: sdk/src/compute_budget_processor.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
+//! Process compute_budget instructions outside of program-runtime to
```
Comment (taozhu-chicago) [inline sdk/src/compute_budget_processor.rs:1]: mostly just moved `process_instructions(...)` from program-runtime to sdk to reduce dependency.

File: sdk/src/transaction/sanitized.rs [hunk 6]
```diff
@@ -67,6 +72,7 @@ impl SanitizedTransaction {
         address_loader: impl AddressLoader,
     ) -> Result<Self> {
         let signatures = tx.signatures;
+        let transaction_meta = tx.transaction_meta;
```
Comment (taozhu-chicago) [inline sdk/src/transaction/sanitized.rs:75]: `transaction_meta` is moved over if `SanitizedTransaction` is created from `SanitizedVersionedTransaction`

File: sdk/src/transaction/sanitized.rs [hunk 7]
```diff
@@ -128,14 +136,23 @@ impl SanitizedTransaction {
             }
         });
 
+        // extract transaction meta
+        let transaction_meta = process_compute_budget_instruction(
+            message.program_instructions_iter(),
+            feature_set,
+            None, // cluster info
+        )?;
+
```
Comment (taozhu-chicago) [inline sdk/src/transaction/sanitized.rs:145]: If creating from `VersionedTransaction`, then process compute_budget instructions once here.

File: sdk/src/transaction/versioned/sanitized.rs [hunk 8]
```diff
@@ -11,27 +14,36 @@ pub struct SanitizedVersionedTransaction {
...
+        tx: VersionedTransaction,
+        feature_set: &FeatureSet,
+    ) -> Result<Self, SanitizeError> {
         tx.sanitize_signatures()?;
+        let sanitized_versioned_message = SanitizedVersionedMessage::try_from(tx.message)?;
+        let transaction_meta = process_compute_budget_instruction(
+            sanitized_versioned_message.program_instructions_iter(),
+            feature_set,
+            None, //cluster type
+        )?;
```
Comment (taozhu-chicago) [inline sdk/src/transaction/versioned/sanitized.rs:32]: process compute_budget instructions once when sanitizing VersionedTransaction, transaction_meta will be moved to SanitizedTransaction.

**Review by taozhu-chicago — COMMENTED** [review]
this draft propose one way to simplify compute_budget instructions processing, to consolidate priority fee and cu-limit related code into one place - transaction meta. 

Initial bench tests shows slight TPS improvement, tho the main gain is reduced code complexity. Would like to get your thoughts on the approach before making it review-able PR(s), and update tests.

CC @t-nelson  @apfitzge

**Comment by apfitzge** [comment]
Just looked at a very high level.

I don't particularly like the movement of **logic** into the `sdk` crate.
The changes we're making here are effectively an optimization on **internal** types, to store meta data at an earlier stage.

I think a as an initial step, we might consider moving `Sanitized(Versioned)Transaction` out of the `sdk` and into `runtime`; afaik these types are only used by `runtime` and `core`.
Then the logic for populating `TransactionMeta` can also go into runtime, which I think makes a lot more sense.

^ Added benefit on this is that it makes it much clearer that `Sanitized(Versioned)Transaction` are internal types that we are free to change the representation of...I've actually been hesistant in the past to change `SanitizedTransaction` because I thought there would be some `sdk` stuff it broke

**Comment by taozhu-chicago** [comment]
> I think as an initial step, we might consider moving Sanitized(Versioned)Transaction out of the sdk and into runtime;

This sounds reasonable, `sanitized` transactions are good candidates of being *internal* type. Not sure if there were reasons they are in `sdk`, @t-nelson might know? 

> I've actually been hesistant in the past to change SanitizedTransaction because I thought there would be some sdk stuff it broke

me too. Not sure what might break by changing 
```
impl SanitizedTransaction{
    pub fn try_create(
        tx: VersionedTransaction,
        message_hash: impl Into<MessageHash>,
        is_simple_vote_tx: Option<bool>,
        address_loader: impl AddressLoader,
        feature_set: &FeatureSet,      <==== additional param
    ) -> Result<Self> {...}
}
```
Ask @t-nelson again 😃

**Comment by t-nelson** [comment]
hmm... thought we removed the ability to deser immediately to Transaction, instead forcing it through SanitizedTransaction but looks like there's no such restriction :thinking: 

imo we went the wrong way with `SantitizedTransaction`.  should've instead introduced an `UnsanitizedTransaction` which implements `Deserialize`, has no accessors for unsanitized fields and can be fallibly converted to `Transaction`. then removed `Deserialize` from `Transaction`.

tl;dr, i think we need them in sdk to do client-side deserialization

File: sdk/src/simple_vote_transaction_checker.rs [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+    instructions_count: usize,
+    is_legacy_message: bool,
+    mut instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+) -> bool {
+    if signatures_count < 3 && instructions_count == 1 && is_legacy_message {
+        instructions.next().map(|(program_id, _ix)| program_id) == Some(&crate::vote::program::id())
+    } else {
+        false
+    }
+}
```
Comment (apfitzge) [inline sdk/src/simple_vote_transaction_checker.rs:14]: I think `instructions_count` is redundant w/ the instructions iterator:


```rust
    if signatures_count < 3 && is_legacy_message {
        let first_ix = instructions.next();
        let next_ix = instructions.next();
        if next_is.is_none() {
           first_ix.map(|(program_id, _ix)| program_id) == Some(&crate::vote::program::id())
        } else {
			false
        }
    } else {
        false
    }
```

alternatively, using `Option::xor`:
```rust
	signatures_count < 3 && is_legacy_message &&
	    instructions.next()
	        .xor(instructions.next())
	        .map(|(program_id, _ix)| program_id == &crate::vote::program::id())
	        .unwrap_or_false()
```

`Option::xor` and how iterators work let's us guarantee that the the first instruction exclusive!

File: sdk/src/compute_budget_processor.rs [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+// Processing compute_budget could be part of tx sanitizing, failed to process
+// these instructions will drop the transaction eventually without execution,
+// may as well fail it early.
+// If succeeded, the transaction's specific limits/requests (could be default)
+// are retrieved and returned,
+pub fn process_compute_budget_instruction<'a>(
+    instructions: impl Iterator<Item = (&'a Pubkey, &'a CompiledInstruction)>,
+    feature_set: &FeatureSet,
+    maybe_cluster_type: Option<ClusterType>,
+) -> Result<TransactionMeta, SanitizeError> {
```
Comment (apfitzge) [inline sdk/src/compute_budget_processor.rs:30]: it seems feasible that `TransactionMeta` will include more than just compute budget information in the future.

I wonder if we'd be better of having some sort of top-level fn for creating the `TransactionMeta`:

```rust
fn new_transaction_meta(...) -> Result<TransactionMeta, SanitizeError> {
    let mut meta = TransactionMeta::default();
    process_compute_budget_instrunction(..., &mut meta)?;
    check_is_simple_vote(..., &mut meta)?;
    ...
    meta
}
```

Comment (apfitzge) [reply sdk/src/compute_budget_processor.rs:30]: ^ on the fence about this exact structure. Potentially better to just have each internal fn return the parts needed, and then build the `Transactionmeta` at the end.

File: core/src/banking_stage/immutable_deserialized_packet.rs [hunk 11]
```diff
@@ -42,33 +39,45 @@ pub struct ImmutableDeserializedPacket {
...
 }
 
 impl ImmutableDeserializedPacket {
     pub fn new(packet: Packet) -> Result<Self, DeserializedPacketError> {
         let versioned_transaction: VersionedTransaction = packet.deserialize_slice(..)?;
-        let sanitized_transaction = SanitizedVersionedTransaction::try_from(versioned_transaction)?;
+        let is_simple_vote = packet.meta().is_simple_vote_tx();
+        // drop transaction if prioritization fails.
+        //
+        // TODO - wire up feature_set from recent bank. For now, mimic current implementation of
```
Comment (apfitzge) [inline core/src/banking_stage/immutable_deserialized_packet.rs:52]: potential issues around feature activations. If we use a recent (rooted?) bank, we likely won't be able to **actually** use features until several slots after the epoch boundary.

File: sdk/src/simple_vote_transaction_checker.rs (hunk 9 above)
Comment (taozhu-chicago) [reply sdk/src/simple_vote_transaction_checker.rs:14]: 🆒 the usage of `option::xor` over iterator! 

But just for the sake of debate, `iterator` itself does not contains `size` information, we could ofc _try_ to extract that info as demo above, but not necessarily reliable. For example, the iter parametyer could be the last iter of !vec(ix; 3)...

With that said, this function does look ugly as it is rn.

File: sdk/src/compute_budget_processor.rs (hunk 10 above)
Comment (taozhu-chicago) [reply sdk/src/compute_budget_processor.rs:30]: totally. Ideally compute_budget_processor shouldn't need to know `TransactionMeta` at all. It could just return a tuple, or a struct?

File: core/src/banking_stage/immutable_deserialized_packet.rs (hunk 11 above)
Comment (taozhu-chicago) [reply core/src/banking_stage/immutable_deserialized_packet.rs:52]: current code uses hard coded feature status for immutable_deserialzied_packet, which this code mimics. Perhaps this is a better/stable option compare to use highest_slot from bank_fork to get feature_set.

File: sdk/src/simple_vote_transaction_checker.rs (hunk 9 above)
Comment (apfitzge) [reply sdk/src/simple_vote_transaction_checker.rs:14]: > For example, the iter parametyer could be the last iter of !vec(ix; 3)...

Yeah, possible but less likely imo. My main point was that we've got 2 args which give us very related info. Easier to screw up when 2 args need consistency

Comment (taozhu-chicago) [reply sdk/src/simple_vote_transaction_checker.rs:14]: Perhaps it can also be argued that although these 2 arg may derived from same vector, but they don't have to be related, especially from this function perspective - it wants to have 1) the size of the vector, 2) iterator of the vector. If can have an iter that is `head` of vector, then that one arg can provide both info.

**Comment by taozhu-chicago** [comment]
to recap discussion at discord: https://discord.com/channels/428295358100013066/439194979856809985/1154833349341352038

1. "transaction sanitize" should be a self-contained task only to ensure the bytes sanely map to a protocol-conforming transaction. Must keep it that way
2. Loaded accounts from ALT, and limits parsed from compute_budget instructions etc are more "internal" data that belongs to `runtime` than `sdk`.
3. suggested approach: leave transactions in `sdk` unchanged, build `runtime` transaction type to encapsulate variable "internal" data

**Comment by taozhu-chicago** [comment]
Going to break up this PR to smaller/reviewable pieces, in the direction of: 
1. leaving `sdk` transaction types unchanged, so `sanitizing` does it purest form of work there.
2. adding transaction types to `runtime` that does address loading and compute budget limits processing, on top of sanitized versioned transaction type from `sdk`. 
 
Identifying a good name is rather challenging, going to start with Andrew's `RuntimeTransaction` first, open for suggestions 😃
